File: ployglots/mta-ployglot-pkg-fetch-client/packages/ts/examples/api/03-event-hooks.ts
// @ts-nocheck
/**
 * Example: event hooks for request-id propagation + response timing.
 *
 * Hooks fire in registration order; a hook that throws aborts the chain and
 * surfaces to the caller. Both `onRequest` and `onResponse` accept a single
 * function or an array of functions.
 */
import { AsyncClient, type Request, type Response } from '@polyglot/fetch-http-client';
import { randomUUID } from 'node:crypto';

async function main(): Promise<void> {
  const startTimes = new WeakMap<Request, number>();

  const client = new AsyncClient({
    baseUrl: 'https://jsonplaceholder.typicode.com',
    eventHooks: {
      onRequest: [
        (req: Request) => {
          const id = randomUUID();
          req.headers.set('x-request-id', id);
          startTimes.set(req, Date.now());
          console.log('→', req.method, req.urlString, 'id=' + id);
        },
      ],
      onResponse: [
        (resp: Response) => {
          const start = resp.request ? startTimes.get(resp.request) : undefined;
          const elapsedMs = start !== undefined ? Date.now() - start : undefined;
          console.log(
            '←',
            resp.statusCode,
            elapsedMs !== undefined ? `in ${elapsedMs}ms` : '(elapsed unknown)',
          );
        },
      ],
    },
  });
  try {
    await client.get('/todos/1');
  } finally {
    await client.close();
  }
}

main().catch((err: unknown) => {
  console.error(err);
  process.exitCode = 1;
});

File: ployglots/mta-ployglot-pkg-fetch-client/packages/py/examples/api/03_event_hooks.py
"""API example: on_request / on_response hooks for request-id + timing."""

from __future__ import annotations

import asyncio
import sys
import time
import uuid

import httpx

from fetch_http_client import AsyncClient


async def main() -> None:
    request_started: dict[str, float] = {}

    async def attach_request_id(req: httpx.Request) -> None:
        req.headers["x-request-id"] = str(uuid.uuid4())
        request_started[req.headers["x-request-id"]] = time.monotonic()

    async def log_response(resp: httpx.Response) -> None:
        rid = resp.request.headers.get("x-request-id", "?")
        started = request_started.pop(rid, time.monotonic())
        sys.stdout.write(
            f"[{rid}] {resp.request.method} {resp.request.url} "
            f"→ {resp.status_code} in {int((time.monotonic() - started) * 1000)}ms\n"
        )

    async with AsyncClient(
        base_url="https://httpbin.org",
        on_request=[attach_request_id],
        on_response=[log_response],
    ) as client:
        await client.get("/get")


if __name__ == "__main__":
    asyncio.run(main())
